### ISSUE TITLE
Enhancing OutputPort Backpressure Handling via RecvError::Lagged Management and Buffer Configurability

### ISSUE THREAD

**Issue by dcadenas** [slawlor/ractor#225]
I've noticed some silent drops on the receiver side under heavy load, mainly because there's no handling for `RecvError::Lagged` in [this section](https://github.com/slawlor/ractor/blob/130bb1a01dd899a8976cf5d5014595aca278db77/ractor/src/port/output/mod.rs#L142). Adjusting the broadcast buffer size could help temporarily, but directly addressing backpressure could offer a more durable solution.

### Dealing with `RecvError::Lagged`:
Implementing explicit handling for `RecvError::Lagged` when backpressure interrupts the recv loop could greatly improve our insight into and control over backpressure issues. Logging or aborting cleanly in these cases would make such events more transparent.

### Using `Sender::len` for Proactive Management:
Checking the channel's capacity with [`Sender::len`](https://docs.rs/tokio/latest/tokio/sync/broadcast/struct.Sender.html#method.len) before sending allows for preemptive action to avoid overloading. This approach could enable smarter decision-making on sending operations, based on the current state of the channel.

### Configurable Broadcast Buffer Size:
Allowing the buffer size to be adjusted externally could provide the flexibility needed to optimize for varying workloads, helping to manage backpressure more effectively.

Integrating these suggestions could enhance how backpressure is managed, leading to a system that’s more robust and easier to fine-tune.

**Comment by slawlor** [comment]
I just wanted to state I'm seeing these posts, and am thrilled with the progress. However a sick 5-month-old has prevented me from doing proper reviews here. Rest assured I will get to it, sorry for the delay

**Comment by dcadenas** [comment]
No worries at all! Family always comes first, and I hope your little one gets well soon. I'm just grateful for the work you've put into the library and am looking forward to your insights whenever you have the time. Take care and thank you for the update!

**Comment by slawlor** [comment]
OK now I've actually had 20 minutes to read this!

I'm actually surprised this is an issue. In the `OutputPort`, the spawned task is just doing a `recv() -> map() -> send()` to the target actor, which has an unbounded buffer. Do you have a repro you can share of where this is missing messages under heavy load? (I understand if it's proprietary or something of course).

However if this is indeed a problem, it means the tokio scheduler is **really** busy that it can't do a simple dequeue of a port, then two blocking operations.

My only concern with pushing logic into the sender space, is that we effect QoS for all downstream targets for a single slow target, which kind of breaks the point of pub-sub models. Probably the correct approach (imho) is dealing with the `LAGGED` signal as you suggest.

**Comment by dcadenas** [comment]
> My only concern with pushing logic into the sender space, is that we effect QoS for all downstream targets for a single slow target, which kind of breaks the point of pub-sub models. Probably the correct approach (imho) is dealing with the LAGGED signal as you suggest.

I hadn't considered that, which makes a lot of sense. Here's some code to reproduce the issue:

```rust
use std::time::Duration;

use tracing_subscriber::{fmt, prelude::*, EnvFilter};

use ractor::{Actor, ActorProcessingErr, ActorRef, OutputPort};

struct Counter;

struct CounterState {
    count: i64,
    output_port: OutputPort<i64>,
}

enum CounterMessage {
    Increment(i64),
    Subscribe(ActorRef<CounterMessage>),
    DoSomethingSlow(i64),
}

#[ractor::async_trait]
impl Actor for Counter {
    type Msg = CounterMessage;

    type State = CounterState;
    type Arguments = ();

    async fn pre_start(
        &self,
        _myself: ActorRef<Self::Msg>,
        _: (),
    ) -> Result<Self::State, ActorProcessingErr> {
        tracing::info!("Starting the counter actor");
        // create the initial state
        Ok(CounterState {
            count: 0,
            output_port: OutputPort::default(),
        })
    }

    async fn handle(
        &self,
        _myself: ActorRef<Self::Msg>,
        message: Self::Msg,
        state: &mut Self::State,
    ) -> Result<(), ActorProcessingErr> {
        match message {
            CounterMessage::Increment(how_much) => {
                if state.count < 5 {
                    tokio::time::sleep(std::time::Duration::from_secs(1)).await;
                }

                if state.count == 5 {
                    tracing::info!("Producer speed up, to ensure that we fill the 10 items broadcast buffer, receiver loop will be broken from lagging messages and silently drop");
                }

                state.count += how_much;
                tracing::info!("Producer current count: {}", state.count);
                state.output_port.send(how_much);
            }
            CounterMessage::Subscribe(subscriber) => {
                state
                    .output_port
                    .subscribe(subscriber, |i64| Some(CounterMessage::DoSomethingSlow(i64)));
            }
            CounterMessage::DoSomethingSlow(value) => {
                tokio::time::sleep(std::time::Duration::from_secs(2)).await;
                state.count += value;
                tracing::info!("Consumer current count: {}", state.count);
            }
        }
        Ok(())
    }
}

fn init_logging() {
    tracing_subscriber::registry()
        .with(fmt::layer())
        .with(EnvFilter::from_default_env())
        .init();
}

#[tokio::main]
async fn main() {
    init_logging();

    let (actor, handle) = Actor::spawn(Some("test_name".to_string()), Counter, ())
        .await
        .expect("Failed to start actor!");

    let (actor_subscriber, handle_subscriber) =
        Actor::spawn(Some("test_subscriber".to_string()), Counter, ())
            .await
            .expect("Failed to start actor!");

    actor
        .send_message(CounterMessage::Subscribe(actor_subscriber.clone()))
        .expect("Failed to send message");

    // +5 +10 -5 a few times, printing the value via RPC
    for _i in 0..25 {
        actor
            .send_message(CounterMessage::Increment(1))
            .expect("Failed to send message");
    }

    tokio::spawn(async move {
        tokio::time::sleep(Duration::from_secs(25)).await;
        actor.stop(None);
        actor_subscriber.stop(None);
    });

    handle.await.expect("Actor failed to exit cleanly");
    handle_subscriber
        .await
        .expect("Actor failed to exit cleanly");
}
```
I'm happy to share the real use case because it's open source. It's a [server](https://github.com/planetary-social/reportinator_server) that consumes direct messages from the [Nostr network](https://nostr.org/).

The server processes messages that can arrive very quickly from a websocket port connected to the Nostr network relays. Initially, I encountered no issues because the test messages from my local test server were fewer than 10. However, as soon as I exceeded that limit, the initial query returned too much data too quickly, so my message handler [here](https://github.com/planetary-social/reportinator_server/blob/49bf378f455a9325dec26c575ec167732f4c13fc/src/actors/event_enqueuer.rs#L54) wasn't being called due to a broken loop. I was puzzled because my subscriber wasn't being called. Eventually, I "solved" it by adding a limit to the initial query [here](https://github.com/planetary-social/reportinator_server/blob/49bf378f455a9325dec26c575ec167732f4c13fc/src/main.rs#L42). This workaround mostly prevents such initial large bursts, which would be rare for the moment, but it's not a real long term solution.

I'm pretty new to Rust and the actor model, so any feedback or suggestion on how to use the crate or anything else will always be welcome.

**Comment by slawlor** [comment]
I'm trying to understand what's going on here, we're creating a broadcast channel with a "buffer" of 10 messages. We were exiting the subscription if we got a `Lagged(num_dropped)` error which isn't right I think. In the event we're lagged, we should log it and probably try and continue on.

However I'm honestly surprised this is happening, because we spawn a dedicated task to take the output port's received messages and forward them to an actor's (unbounded) channel. The only thing I could think of is if the converter is really slow converting messages from type A -> B.

Like as long as the subscription is running, we shouldn't ever fill the buffer of 10 unless we're under super heavy cpu pressure I would hope.

**Comment by slawlor** [comment]
I added a branch to handle the lagged error, but i'm not in love with the idea of dropping messages. It makes me wonder if I should write my own implementation of the broadcast channel...

https://github.com/slawlor/ractor/commit/d4767b2c7435535e6164401b0126a4ca5a8f2807

**Comment by philjb** [comment]
tl;dr: I repeat @dcadenas 
> Adjusting the broadcast buffer size could help temporarily, but directly addressing backpressure could offer a more durable solution.

We at InfluxDB have also encountered this issue's situation, nearly identically as @dcadenas and @daveman1010221 have described. 

In brief, for us, we have a separate process make an rpc call with a prefix to our ractor based service. That prefix is compared against internal book keeping to generate messages on an `OutputPort`. Sometimes that prefix matches only a few internal items generating only a few messages (it's 1:1); sometimes it is hundreds or more. Our dispatch looks roughly like this: where it calls `send` on the output port in a non-async loop. 

```
// internal_state is a HashMap
// subscribers is a OutputPort<Event>; Event is an enum
state.internal_state.retain(|item| match prefix {
    if prefix.is_match(item) => {
        state
            .subscribers
            .send(Event::Match(item.clone());
        false
    } else {
         true
    }
});
```

We can reliably encounter the Lagged situation by setting it up such that the retain loop sends a 100 messages or so. I can replicate this on my M2 Max 64GB machine. I have not maxed out cpu in doing so. 

I believe it is because the tokio broadcast implementation, `send`ing is synchronous so I can send extremely fast on one task and it doesn't matter much if there is capacity on other workers. The "forwarding" task internal to ractor has to be woken for each value (although I know tokio has optimization for seeing if the same task is ready already to be woken again): We're on 0.14.17.

In any case, several reports have shown this can happen. For me, the issue is this makes `Output` ports an unreliable feature of ractor. If they are advertised as best effort delivery, I wouldn't have used them since we can't miss messages for correctness. Worse though is that the internal forwarding task exits silently and all forwarding stops with the current implementation here: 

https://github.com/slawlor/ractor/blob/f9ccc48d16dcddb9ff87f55987e2089bfb9fbb47/ractor/src/port/output/mod.rs#L130-L137

It is an unfortunate api for library users that it is a silent, undocumented failure/risk; there's not currently a way for our code to detect this situation (in the OutputPort's apis) and remedy without a process restart. If I had known this was a risk with `OutputPort`s, we would have mitigated originally as we are doing now: which is a combination of batching our information into one message (instead of 100s), adding delays in the synchronous sending, and forking ractor to configure the capacity (we'll upstream a pr shortly for that), or not using OutputPorts (although they are a nice feature to include in the library).

This is a result of tokio's tradeoffs for broadcast channels. If I'm using a broadcast channel directly, I'm aware of the tradeoffs and forced to handle (or ignore) the Err<Lagging> situation. Ractor breaks as sends can still use the output port but don't know receivers will never get the message; and receivers never know no more messages will ever arrive.

> Perhaps some custom options supported on the parent structure which can let the user define behavior (i.e. is dropping messages ok?) would be good.

I think [your branch](https://github.com/slawlor/ractor/commit/d4767b2c7435535e6164401b0126a4ca5a8f2807) for handling Lagging is a way: but at least the receiving actor should get an error or notice that it missed a message as then our code can make a determination on how to handle that situation. 
 
I understand why `10` was picked as the channel capacity as broadcast channels do a Vec<type> allocation of the capacity. I think exposing an api in ractor so that library users could determine the tradeoff of memory and lagging risk for themselves helps (as @dcadenas suggested in OP). We'll upstream a PR soon.

I'm not sure [Sender::len](https://docs.rs/tokio/latest/tokio/sync/broadcast/struct.Sender.html#method.len) helps though as I think that is inherently racy with multiple concurrent senders possible.

Then I think something has to be done about the forwarding task: 

https://github.com/slawlor/ractor/blob/f9ccc48d16dcddb9ff87f55987e2089bfb9fbb47/ractor/src/port/output/mod.rs#L130-L137

Perhaps making that an actor itself that can fail and is supervised (similar to ractor's design for stream pumps).  At the least, I would want to be able to know when Lagging occurred so I could have our process decide what to do (which is maybe terminate with an unrecoverable error). We'll spend some thought on this and hopefully offer a pr option or two.

I have seen but not reviewed in detail the experimental output port v2 option (from Aug 2025): https://github.com/slawlor/ractor/pull/369 for guaranteed delivery, but as you point out it's a `queueing problem [..], which isn't trivial or ideal to solve`.

long post; thank you for reading; we really enjoy ractor so thank you! Generally it's worked great for us.
